File: 백준/C++/단계별/20단계 - 스택/9012번 - 괄호.cpp
#include <iostream>
#include <algorithm>
#include <stack>

using namespace std;

string str;
int t;

string checkVPS(string input)
{
	stack<char> st;
	for (int i = 0; i < input.length(); i++) {
		if (input[i] == '(') {
			st.push(input[i]);
		}
		else {
			if (st.empty()) {
				return "NO";
				break;
			}
			st.pop();
		}
	}

	if (st.empty())
		return "YES";
	else
		return "NO";
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> t;

	while (t--) {
		cin >> str;
		
		cout << checkVPS(str) << "\n";
	}

	return 0;
}	

File: 백준/Python/Silver/9012번 - 괄호.py
import sys

t = int(sys.stdin.readline())

for _ in range(t):
    str = sys.stdin.readline()
    stack = []

    for i in str:
        if i == '(':
            stack.append(i)
        elif i == ')':
            if stack:
                stack.pop()
            else:
                print('NO')
                break 
    else:
        if not stack:
            print('YES')
        else:
            print('NO')